source/develop/release-management/features/virt/engine-dedicated-cpus.md

```diff
@@ -0,0 +1,139 @@
+    * Poll every X seconds all VMs and find which VM’s have cpu pinning and update the shared CPU pinning data structure based on VM pinned CPUs and host capabilities
+
+    Option 2:
+    * On updateVDSStatisticsData update the data structure with host capabilities
+    * On every start/stop/migrate/… VM update the shared CPU pinning data structure
+    With used pinning
+
+   **Init cached data structure (on engine start/stop):**
+   
+    Option1:
+    * Will be calculated after engine started after getting host caps and VMs cpu pinnings
+    
+    Option2:
+    * Create table in DB and save/read it from DB
+
+4. Scheduling
+    1. Start VM:
+        - Scheduling phase
+            - Reads hostToAvailableCpu from step 3
+            - Filter hosts that can't match virtual CPU topology (pCpu pinning should
+              be close to the virtual topology, e.g: 2 vCPU on same vSocket should be placed on same pSocket)
+            - Score all hosts to find best matching host with enough pCPUs 
+        - Run phase
+            - Based on Step 3 -Save/Update cached data structure
+            - Update VM CurrentCpuPinning DB field with the mapping
+
+    2. Stop/Destroy VM:
+        - Clear VM dynamic CPU pinning field
+        - Based on Step 3 -Save/Update cached data structure
+    
+3. Send vcpupin with cpusets and policy name in VM metadata
+   1. Send the policy name, available policy names: { none, pin, dedicated, siblings, isolate-threads }.
+   2. Send the cpusets (it will be stored in VM_Dynamic)
+
+4. Migrate VDSM call
+   1. Calculate vCPU to pCPU mapping based on the scheduled host
+   2. Send cpusets additional parameter
+   
+5. setNumberOfCpus VDSM call
+   1. Send cpusets additional parameter
+
+
+## Database
+
+* Save CPU mappings in field CurrentCpuPinning field in vm_dynamic table
+* Save CPU topology from host capabilities into vds_statistics
+* Save new pinning policies to DB using the existing CpuPinningPolicy field
+
+## Rest API
+
+1. Add new types of CPU pinning to api-model
+2. Update mapper with the new policies
+
+## Frontend
+
+- The added CPU pinning policy fields should be disabled on cluster level version < 4.7
+
+
+### Example
+2 hosts are running and reporting the following available CPU topology
+
+Host A: 1 Socket, 4 CPUs
+
+Host B: 2 Sockets, 1 CPU on each socket
+
+VM A with dedicated cpu pinning is already running on host A and have 2 vCPUs
+
+User wants to start VM B that requires 2 CPUs and has dedicated cpu policy
+
+- hostToAvailableCpu will be:
+  {host A id} -> { {socket 1 id} -> {1,2}}
+  {host B id} -> { {socket 1 id} -> {3}, {socket 2 id} -> {4}}
+- Scheduler ranks all hosts that have enough physical CPUs available based on hostToAvailableCpu. In our case, both
+  hosts have enough pCPUs, however, host A has both available on one socket whereas the B would need to divide them on 
+  2 different sockets. So A gets higher rank than B (B should be filtered)
```

Comment:
what if the VM is set to 2 sockets as well? the host will be filtered out?

Reply:
we should try to see if there is a host that fit the virtual topology 


if there is host with >2 sockets but we can't manage to fit the virtual socket cpus to the physical one(meaning cpus will need to run on separate sockets), then it will be filtered

if  there is host that we can fit then it will not be filtered

Reply:
My question was about this situation:
If i have a VM with 2 sockets, 1 core, 1 thread.
I have a host A with 1 socket, 1 core, 2 threads.
I have a host B with 2 sockets, 1 core, 1 thread.

Which one will you choose?
Or better asking, will you prefer to fit the virtual topology of the VM to the physical? Or rather, looking onto the physical to run on the same socket?

Reply:
I think in this case we should fit the virtual topology

if we can't fit the virtual topology, we can decide between two options:

1- Try to fit as much as we can from virtual topology, for example try to fit at least 1 socket from virtual topology and and if the second socket can't fit, than try to fit CPU's in separate sockets( if there is still sockets available)

2- Throw validation error, since probably the end-user will not gain too much performance from the CPU pinning that he is planning to do

Reply:
this is an open discussion, TBH i'm not sure what is preferable.
1. we can be close as much we can to the virtual topology - best to fit the user's configuration.
2. we can do it by physical to have better performance.
but, without dedicated policy having 2 sockets could be on same physical socket or elsewhere.

Reply:
I wouldn't filter the hosts based on this computation but rather score them differently (e.g., the lower the amount of sockets + cores, the higher the score will be)